releases/_posts/2018-11-05-spark-release-2-4-0.md

```diff
@@ -0,0 +1,188 @@
+   - [[SPARK-25234](https://issues.apache.org/jira/browse/SPARK-25234)] avoid integer overflow in parallelize
+   - [[SPARK-25117](https://issues.apache.org/jira/browse/SPARK-25117)] Add EXCEPT ALL and INTERSECT ALL support in R
+   - [[SPARK-24537](https://issues.apache.org/jira/browse/SPARK-24537)] Add array_remove / array_zip / map_from_arrays / array_distinct
+   - [[SPARK-24187](https://issues.apache.org/jira/browse/SPARK-24187)] Add array_join function to SparkR
+   - [[SPARK-24331](https://issues.apache.org/jira/browse/SPARK-24331)] Adding arrays_overlap, array_repeat, map_entries to SparkR
+   - [[SPARK-24198](https://issues.apache.org/jira/browse/SPARK-24198)] Adding slice function to SparkR
+   - [[SPARK-24197](https://issues.apache.org/jira/browse/SPARK-24197)] Adding array_sort function to SparkR
+   - [[SPARK-24185](https://issues.apache.org/jira/browse/SPARK-24185)] add flatten function to SparkR
+   - [[SPARK-24069](https://issues.apache.org/jira/browse/SPARK-24069)] Add array_min / array_max functions
+   - [[SPARK-24054](https://issues.apache.org/jira/browse/SPARK-24054)] Add array_position function / element_at functions
+   - [[SPARK-23770](https://issues.apache.org/jira/browse/SPARK-23770)] Add repartitionByRange API in SparkR
+
+*Programming guide: <a href="{{site.baseurl}}/docs/2.4.0/sparkr.html">SparkR (R on Spark)</a>.*
+
+
+### GraphX
+
+ - **Major features**
+   - [[SPARK-25268](https://issues.apache.org/jira/browse/SPARK-25268)] run Parallel Personalized PageRank throws serialization Exception
+
+*Programming guide: <a href="{{site.baseurl}}/docs/2.4.0/graphx-programming-guide.html">GraphX Programming Guide</a>.*
+
+
+### Deprecations
+
+ - **MLlib**
+   - [[SPARK-23451](https://issues.apache.org/jira/browse/SPARK-23451)] Deprecate KMeans computeCost
+   - [[SPARK-25345](https://issues.apache.org/jira/browse/SPARK-25345)] Deprecate readImages APIs from ImageSchema
+
+
+### Changes of behavior
+
+ - **Spark Core**
+   - [[SPARK-25088](https://issues.apache.org/jira/browse/SPARK-25088)] Rest Server default & doc updates
+
+ - **Spark SQL**
+   - [[SPARK-23549](https://issues.apache.org/jira/browse/SPARK-23549)] Cast to timestamp when comparing timestamp with date
+   - [[SPARK-24324](https://issues.apache.org/jira/browse/SPARK-24324)] Pandas Grouped Map UDF should assign result columns by name
+   - [[SPARK-23425](https://issues.apache.org/jira/browse/SPARK-23425)] load data for hdfs file path with wildcard usage is not working properly
+   - [[SPARK-23173](https://issues.apache.org/jira/browse/SPARK-23173)] from_json can produce nulls for fields which are marked as non-nullable
+   - [[SPARK-24966](https://issues.apache.org/jira/browse/SPARK-24966)] Implement precedence rules for set operations
+   - [[SPARK-25708](https://issues.apache.org/jira/browse/SPARK-25708)] HAVING without GROUP BY should be global aggregate
```

Comment:
In the migration guide, we have more contents than the list of these JIRAs, right?
```
  - Since Spark 2.4, when there is a struct field in front of the IN operator before a subquery, the inner query must contain a struct field as well. In previous versions, instead, the fields of the struct were compared to the output of the inner query. Eg. if `a` is a `struct(a string, b int)`, in Spark 2.4 `a in (select (1 as a, 'a' as b) from range(1))` is a valid query, while `a in (select 1, 'a' from range(1))` is not. In previous version it was the opposite.

  - In versions 2.2.1+ and 2.3, if `spark.sql.caseSensitive` is set to true, then the `CURRENT_DATE` and `CURRENT_TIMESTAMP` functions incorrectly became case-sensitive and would resolve to columns (unless typed in lower case). In Spark 2.4 this has been fixed and the functions are no longer case-sensitive.

  - Since Spark 2.4, Spark will evaluate the set operations referenced in a query by following a precedence rule as per the SQL standard. If the order is not specified by parentheses, set operations are performed from left to right with the exception that all INTERSECT operations are performed before any UNION, EXCEPT or MINUS operations. The old behaviour of giving equal precedence to all the set operations are preserved under a newly added configuration `spark.sql.legacy.setopsPrecedence.enabled` with a default value of `false`. When this property is set to `true`, spark will evaluate the set operators from left to right as they appear in the query given no explicit ordering is enforced by usage of parenthesis.

  - Since Spark 2.4, Spark will display table description column Last Access value as UNKNOWN when the value was Jan 01 1970.

  - Since Spark 2.4, Spark maximizes the usage of a vectorized ORC reader for ORC files by default. To do that, `spark.sql.orc.impl` and `spark.sql.orc.filterPushdown` change their default values to `native` and `true` respectively.

  - In PySpark, when Arrow optimization is enabled, previously `toPandas` just failed when Arrow optimization is unable to be used whereas `createDataFrame` from Pandas DataFrame allowed the fallback to non-optimization. Now, both `toPandas` and `createDataFrame` from Pandas DataFrame allow the fallback by default, which can be switched off by `spark.sql.execution.arrow.fallback.enabled`.

  - Since Spark 2.4, writing an empty dataframe to a directory launches at least one write task, even if physically the dataframe has no partition. This introduces a small behavior change that for self-describing file formats like Parquet and Orc, Spark creates a metadata-only file in the target directory when writing a 0-partition dataframe, so that schema inference can still work if users read that directory later. The new behavior is more reasonable and more consistent regarding writing empty dataframe.

  - Since Spark 2.4, expression IDs in UDF arguments do not appear in column names. For example, a column name in Spark 2.4 is not `UDF:f(col0 AS colA#28)` but ``UDF:f(col0 AS `colA`)``.

  - Since Spark 2.4, writing a dataframe with an empty or nested empty schema using any file formats (parquet, orc, json, text, csv etc.) is not allowed. An exception is thrown when attempting to write dataframes with empty schema.

  - Since Spark 2.4, Spark compares a DATE type with a TIMESTAMP type after promotes both sides to TIMESTAMP. To set `false` to `spark.sql.legacy.compareDateTimestampInTimestamp` restores the previous behavior. This option will be removed in Spark 3.0.

  - Since Spark 2.4, creating a managed table with nonempty location is not allowed. An exception is thrown when attempting to create a managed table with nonempty location. To set `true` to `spark.sql.legacy.allowCreatingManagedTableUsingNonemptyLocation` restores the previous behavior. This option will be removed in Spark 3.0.

  - Since Spark 2.4, renaming a managed table to existing location is not allowed. An exception is thrown when attempting to rename a managed table to existing location.

  - Since Spark 2.4, the type coercion rules can automatically promote the argument types of the variadic SQL functions (e.g., IN/COALESCE) to the widest common type, no matter how the input arguments order. In prior Spark versions, the promotion could fail in some specific orders (e.g., TimestampType, IntegerType and StringType) and throw an exception.

  - Since Spark 2.4, Spark has enabled non-cascading SQL cache invalidation in addition to the traditional cache invalidation mechanism. The non-cascading cache invalidation mechanism allows users to remove a cache without impacting its dependent caches. This new cache invalidation mechanism is used in scenarios where the data of the cache to be removed is still valid, e.g., calling unpersist() on a Dataset, or dropping a temporary view. This allows users to free up memory and keep the desired caches valid at the same time.

  - In version 2.3 and earlier, Spark converts Parquet Hive tables by default but ignores table properties like `TBLPROPERTIES (parquet.compression 'NONE')`. This happens for ORC Hive table properties like `TBLPROPERTIES (orc.compress 'NONE')` in case of `spark.sql.hive.convertMetastoreOrc=true`, too. Since Spark 2.4, Spark respects Parquet/ORC specific table properties while converting Parquet/ORC Hive tables. As an example, `CREATE TABLE t(id int) STORED AS PARQUET TBLPROPERTIES (parquet.compression 'NONE')` would generate Snappy parquet files during insertion in Spark 2.3, and in Spark 2.4, the result would be uncompressed parquet files.

  - Since Spark 2.0, Spark converts Parquet Hive tables by default for better performance. Since Spark 2.4, Spark converts ORC Hive tables by default, too. It means Spark uses its own ORC support by default instead of Hive SerDe. As an example, `CREATE TABLE t(id int) STORED AS ORC` would be handled with Hive SerDe in Spark 2.3, and in Spark 2.4, it would be converted into Spark's ORC data source table and ORC vectorization would be applied. To set `false` to `spark.sql.hive.convertMetastoreOrc` restores the previous behavior.

  - In version 2.3 and earlier, CSV rows are considered as malformed if at least one column value in the row is malformed. CSV parser dropped such rows in the DROPMALFORMED mode or outputs an error in the FAILFAST mode. Since Spark 2.4, CSV row is considered as malformed only when it contains malformed column values requested from CSV datasource, other values can be ignored. As an example, CSV file contains the "id,name" header and one row "1234". In Spark 2.4, selection of the id column consists of a row with one column value 1234 but in Spark 2.3 and earlier it is empty in the DROPMALFORMED mode. To restore the previous behavior, set `spark.sql.csv.parser.columnPruning.enabled` to `false`.

  - Since Spark 2.4, File listing for compute statistics is done in parallel by default. This can be disabled by setting `spark.sql.statistics.parallelFileListingInStatsComputation.enabled` to `False`.

  - Since Spark 2.4, Metadata files (e.g. Parquet summary files) and temporary files are not counted as data files when calculating table size during Statistics computation.

  - Since Spark 2.4, empty strings are saved as quoted empty strings `""`. In version 2.3 and earlier, empty strings are equal to `null` values and do not reflect to any characters in saved CSV files. For example, the row of `"a", null, "", 1` was writted as `a,,,1`. Since Spark 2.4, the same row is saved as `a,,"",1`. To restore the previous behavior, set the CSV option `emptyValue` to empty (not quoted) string.  

  - Since Spark 2.4, The LOAD DATA command supports wildcard `?` and `*`, which match any one character, and zero or more characters, respectively. Example: `LOAD DATA INPATH '/tmp/folder*/'` or `LOAD DATA INPATH '/tmp/part-?'`. Special Characters like `space` also now work in paths. Example: `LOAD DATA INPATH '/tmp/folder name/'`.

  - In Spark version 2.3 and earlier, HAVING without GROUP BY is treated as WHERE. This means, `SELECT 1 FROM range(10) HAVING true` is executed as `SELECT 1 FROM range(10) WHERE true`  and returns 10 rows. This violates SQL standard, and has been fixed in Spark 2.4. Since Spark 2.4, HAVING without GROUP BY is treated as a global aggregate, which means `SELECT 1 FROM range(10) HAVING true` will return only one row. To restore the previous behavior, set `spark.sql.legacy.parser.havingWithoutGroupByAsWhere` to `true`.
```

Reply:
thanks for pointing it out! I think we don't need to mention all of them in release notes, like we don't mention all the resolved tickets. I'll pick some important ones.